### ISSUE TITLE
 TLS handshake error from API server

### ISSUE THREAD

**Issue by sknmi** [aws/karpenter-provider-aws#6898]
### Description

**Observed Behavior**:
```
karpenter-c595bb5d8-8r8jr controller {"level":"ERROR","time":"2024-08-30T08:06:16.304Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:40666: EOF\n","commit":"62a726c"}
karpenter-c595bb5d8-hzfgs controller {"level":"ERROR","time":"2024-08-30T08:07:18.550Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:58290: EOF\n","commit":"62a726c"}
karpenter-c595bb5d8-8r8jr controller {"level":"ERROR","time":"2024-08-30T08:07:18.571Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:55794: EOF\n","commit":"62a726c"}
karpenter-c595bb5d8-8r8jr controller {"level":"ERROR","time":"2024-08-30T08:07:18.572Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:55792: EOF\n","commit":"62a726c"}
karpenter-c595bb5d8-hzfgs controller {"level":"ERROR","time":"2024-08-30T08:08:10.419Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:43424: EOF\n","commit":"62a726c"}
karpenter-c595bb5d8-8r8jr controller {"level":"ERROR","time":"2024-08-30T08:08:10.427Z","logger":"webhook","message":"http: TLS handshake error from 10.x.x.x:52314: EOF\n","commit":"62a726c"}
```
**Expected Behavior**:
No errors :)
**Reproduction Steps** (Please include YAML):
Karpenter on fargate in karpenter namespace.
These messages started to appear after upgrading to 1.0.1
**Versions**:
- Chart Version: 1.0.1
- Kubernetes Version (`kubectl version`): 1.30

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by levinedaniel** [comment]
I don't think this issue should be closed. I am seeing a similar error in my log messages and require the webhook to remain enabled to facilitate the conversion to the latest api version for my resources.

**Comment by ezh** [comment]
I agree with @levinedaniel. What is the reason to mark solution as closed with
```
webhook:
  enabled: false
```
The webhook is broken.

**Comment by Hronom** [comment]
Same

**Comment by Hronom** [comment]
cc @sknmi message above

**Comment by m0untains** [comment]
Also seeing this issue after upgrading to `v0.37.3`.

**Comment by adawalli** [comment]
Saw this issue on `0.37.3` and `1.0.1`

**Comment by liafizan** [comment]
Here is my observation. Please let me know if this is incorrect: 

Karpenter does not provide a ca-client bundle as we can see from [here](https://github.com/aws/karpenter-provider-aws/blob/v1.0.0/charts/karpenter-crd/templates/karpenter.sh_nodepools.yaml#L1084-L1096).

When I look at the CRD in my cluster, I can see that it has been injected with a caBundle:

```
 webhook:
      clientConfig:
        caBundle: Redacted...
        service:
          name: karpenter
          namespace: karpenter
          path: /conversion/karpenter.sh
          port: 8443
      conversionReviewVersions:
      - v1beta1
      - v1
  group: karpenter.sh
```
I believe this is happening through [ca-injector](https://cert-manager.io/docs/concepts/ca-injector/). So this means, that client config for this webhook has a ca-bundle specified but karpenter uses knative to inject certificate data into `karpernter-cert`secret which comes from [here](https://github.com/knative/pkg/blob/28b58b842f1a59212ba919f862e2a030c41a6368/webhook/certificates/resources/certs.go#L144). 

So this means that CA for CRD & Webhooks do not match and hence the error. If this is correct, then may be we can look at the possible solutions

**Comment by jmdeal** [comment]
This appears to be the same issue we saw with the our defaulting / validating webhooks previously, the original issue was closed out when those webhooks were disabled by default: https://github.com/kubernetes-sigs/karpenter/issues/718. I've been able to reproduce, and as with that issues there does not appear to be any actual impact to Karpenter's operation and the errors can be safely ignored.

From the original issue: 
> These TLS errors appear to be related to https://github.com/kubernetes/kubernetes/issues/109022 which states that these handshake errors may be generated by some caching mechanism that is happening in the standard library that causes TLS errors on a cert rotation.

@liafizan are you still running into this? The cert is injected by knative, and I've been unable to reproduce. If you're still encountering this, I'd recommend opening a separate issue. I don't think it's related to the TLS errors we're seeing here.
> I am still not sure how CA bundle is injected in CRD and I did see at one point that the CA bundle in secret vs CRD was different.

**Comment by laserpedro** [comment]
Hello @jmdeal, 

After upgrading to minor 0.37.5 to enable the deleting of webhooks when deployed with ArgoCD I see two things:
- first the validating and mutating webhooks are now properly deleted using ArgoCD.
- the second one is that my CRDs are not in version v1 and are still in v1beta1 so IMO the TLS handshake error is causing the conversion webhook to fail, which is a problem with Karpenter migration to v1.0.x. 
`kubectl get crd nodeclaims.karpenter.sh -o jsonpath='{.spec.versions[*].name}' ` =. v1 v1beta1 / So both versions exist in the cluster.
Therefore the TLS handshake error in my case seems to prevent the validating webhook to perform the v1 migration. I checked the logs inside the controller and that is all I got from the webhook ...

**Comment by laserpedro** [comment]
@jmdeal thank you for your answer, I misunderstood the conversion webhook and thought is was the other way around, thanks for the clarification !

**Comment by elihuj117** [comment]
We are seeing this same behavior. Upgrade from 0.37.0 to 1.0.3 (with a minor upgrade to 0.37.3 during the upgrade process). The error seems to be innocuous, but I wanted to see if there was any impact to the core functionality of Karpenter.

**Comment by apurvabhandari** [comment]
I have done the upgrade from 0.37.5 to 1.0.6 and still see this issue. I have enabled webhook in 0.37.5
`{"level":"ERROR","time":"2024-10-09T14:27:06.147Z","logger":"webhook","message":"http: TLS handshake error from 10.214.2.206:34084: EOF\n","commit":"6174c75"}
{"level":"ERROR","time":"2024-10-09T14:27:06.319Z","logger":"webhook","message":"http: TLS handshake error from 10.214.60.56:40108: EOF\n","commit":"6174c75"}
`

**Comment by itayvolo** [comment]
+1

**Comment by awoimbee** [comment]
> I have done the upgrade from 0.37.5 to 1.0.6 and still see this issue.

I think this issue is caused by the conversion webhook configured on the CRDs (I have had a hard time with these already with #6818).
I use pulumi transforms to remove them, the error is gone:
```typescript
    transforms: [
      ({ props, opts, type }) => {
        if (type === "kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition") {
          // Disable Karpenter conversion webhooks which was only useful when upgrading to v1 and now causes errors
          props.spec.conversion = undefined;
          return { props, opts };
        }
        return undefined;
      }
    ]
```

**Comment by ajith-thomas-fw** [comment]
hi, I did the karpenter version upgrade from v0.33.10 to v1.0.3 following the upgrade guide, https://karpenter.sh/docs/upgrading/v1-migration/#upgrade-procedure,
but as mentioned above by others, ran into the TLS error, but without any impact on the karpenter functionalities.
```
{"level":"ERROR","time":"2024-11-01T05:16:43.587Z","logger":"webhook","message":"http: TLS handshake error from 100.x.x.x:32858: read tcp 100.x.x.x:8443->100.x.x.x:32858: read: connection reset by peer\n","commit":"688ea21"}
{"level":"ERROR","time":"2024-11-01T05:16:43.590Z","logger":"webhook","message":"http: TLS handshake error from 100.x.x.x:32876: read tcp 100.x.x.x:8443->100.x.x.x:32876: read: connection reset by peer\n","commit":"688ea21"}
```
i was able to ignore the errors by disabling the webhook by setting `DISABLE_WEBHOOK=true`. but as mentioned in the below thread, i am also not sure on the repercussions of this. 
https://github.com/kubernetes-sigs/karpenter/issues/718#issuecomment-2447546036

following the discussions in threads, i believe these webhooks are necessary to migrate the api from v1beta1 to v1 in future release. can someone comment on this.

**Comment by Hronom** [comment]
I would like to hear clarifications about this from developers. 
Specifically what is the recommended way if you use latest version of karpenter.

I still don't understand for what webhooks is used for and if I need to keep them enabled in latest version of karpenter.

**Comment by prad9192** [comment]
One of the things I notice is that if we run a single replica of Karpenter, this error goes away. Not a recommendation, but reporting an observation if it helps the investigation.

**Comment by nantiferov** [comment]
I hope this error will be gone with update to 1.1, which should support only v1 API.

**Comment by korncola** [comment]
devs dead or why they are not responding? what are these webhooks, is disabling them safe?
Reminds mind of not using k8s hobbiest products in prod, its just too unreliable to work with.

**Comment by ajaykumarmandapati** [comment]
Issue exists after upgrading to Karpenter v1.1.1 , it is quite misleading and pollutes our logs. Do you'l recommend to turn off the webhook?

`"message":"http: TLS handshake error from [2a05:d014:3b8:5c05::221b]:41978: EOF\n","commit":"a2875e3"}`

**Comment by Hronom** [comment]
How will maintainers know about an issue if it is auto-closed?

**Comment by Hronom** [comment]
ping

**Comment by afonsoaugusto** [comment]
I am also experiencing the same issue.  
I tried upgrading to the latest versions 0.37.6 and 0.36.9, but I still encounter the same problem even after disabling the webhook in the Helm chart for the CRDs.  

However, I recently tried using version 0.36.2, which does not include the webhook, and it has remained stable.  

Nevertheless, I want to be able to upgrade to the latest version, 1.2, and I am awaiting a response to this issue.

**Comment by zhongyi-fu** [comment]
Seems not showing up for me after adding `webhook.serviceNamespace` config for `karpenter-crd` chart. (version: 0.37.7)

**Comment by KRE100** [comment]
I saw the same issues on versions 0.37.7 -> 1.0.8 but I do not see it on version 1.1.3

**Comment by samuel-esp** [comment]
Encountering the same error on version 0.37.6 but i recall this to be present also in previous versions. Usually restarting the deploy solves this. It could be useful to exit the process when the TLS Handshake Error happens too many times consecutively

**Comment by adawalli** [comment]
not stale

**Comment by afonsoaugusto** [comment]
Any updates?

**Comment by jsonmp-k8** [comment]
any updates on this ?

**Comment by leonchik1976** [comment]
problem still exists

**Comment by BernardesPedro** [comment]
problem still exists
V1.0.0
v1.05
v1.0.9
v1.1.5

**Comment by afonsoaugusto** [comment]
The problem is not solved.
Any other issues to follow the evolution?

**Comment by BernardesPedro** [comment]
> The problem is not solved. Any other issues to follow the evolution?

@afonsoaugusto 

For now I managed to solve it by deploying karpenter-crd and disabling the webhook

**Comment by jmdeal** [comment]
Since there's still activity here, I'll link back to my previous comment on this thread: https://github.com/aws/karpenter-provider-aws/issues/6898#issuecomment-2394184893

These errors are spurious and can be safely ignored. They are present for many projects in the Kubernetes ecosystem which utilize webhooks (see https://github.com/kubernetes/kubernetes/issues/109022). The conversion webhooks were only included to facilitate migration from `v1beta1` to `v1` and have been dropped in Karpenter v1.1+. 

Unless there is an actual behavior issue that can be linked to this log, we will not be taking any action to address this. If you believe you have a behavioral issue that's linked to this log, please open a new issue rather than posting here so it's more visible to maintainers. If you just want to get rid of this error log, the recommended path is upgrading to Karpenter v1.1+.

**Comment by SSJNM** [comment]
@BernardesPedro , Thanks it works after deploying karpenter-crd and disabling the webhook. Only disabling webhook didn't worked. 

But webhook conversion should have worked.
